plugins/node/opentelemetry-instrumentation-express/test/hooks.test.ts

```diff
@@ -169,4 +169,140 @@ describe('ExpressInstrumentation hooks', () => {
       );
     });
   });
+
+  describe('span attributes hooks', () => {
+    let server: http.Server;
+    let port: number;
+    let rootSpan: Span;
+
+    beforeEach(async () => {
+      rootSpan = tracer.startSpan('rootSpan');
+
+      const httpServer = await serverWithMiddleware(tracer, rootSpan, app => {
+        app.get('*', (req, res) => {
+          res.send('ok');
+        });
+      });
+      server = httpServer.server;
+      port = httpServer.port;
+    });
+
+    afterEach(() => {
+      server.close();
+    });
+
+    it('should set span attributes', async () => {
+      instrumentation.setConfig({
+        spanAttributesHook: ({ request, route, layerType }, baseAttributes) => {
+          if (layerType) {
+            return {
+              route,
+            };
+          }
+
+          if (route === '*') {
+            return {
+              method: request.method,
+              url: request.url,
+            };
+          }
+
+          return baseAttributes;
+        },
+      });
+
+      await context.with(
+        trace.setSpan(context.active(), rootSpan),
+        async () => {
+          await httpRequest.get(`http://localhost:${port}/foo/3`);
+          rootSpan.end();
+
+          const spans = memoryExporter.getFinishedSpans();
+          assert.strictEqual(spans.length, 2);
+
+          assert.deepEqual(
+            spans.find(span => span.name === 'GET *')?.attributes,
+            { method: 'GET', url: '/foo/3' }
+          );
+
+          assert.deepEqual(
+            spans.find(span => span.name === 'request handler - *')?.attributes,
+            {
+              method: 'GET',
+              url: '/foo/3',
+              'express.name': '*',
+              'express.type': 'request_handler',
+            }
+          );
+        }
+      );
+    });
+
+    it('should use the base attributes when hook throws an error', async () => {
+      instrumentation.setConfig({
+        spanAttributesHook: () => {
+          throw new Error();
+        },
+      });
+
+      await context.with(
+        trace.setSpan(context.active(), rootSpan),
+        async () => {
+          await httpRequest.get(`http://localhost:${port}/foo/3`);
+          rootSpan.end();
+
+          const spans = memoryExporter.getFinishedSpans();
+          assert.strictEqual(spans.length, 2);
+
+          assert.deepEqual(
+            spans.find(span => span.name === 'GET *')?.attributes,
+            {}
+          );
+
+          assert.deepEqual(
+            spans.find(span => span.name === 'request handler - *')?.attributes,
+            {
+              'express.name': '*',
+              'express.type': 'request_handler',
+              'http.route': '*',
+            }
+          );
+        }
+      );
+    });
+
+    it('should use the base attributes when returning undefined from hook', async () => {
+      const spanAttributesHook: SpanAttributesHook = () => {
+        return undefined as unknown as SpanAttributes;
```

Comment:
Looks like implementation supports returning the value `undefined` from the hook to signal no change in attributes. What do you think about adding it to the function return type in `types.ts`. This way we don't have to do these non-idiomatic castings.